### PR TITLE
[Enhancement] Support limit dir size for spill

### PR DESCRIPTION
**Pull request by caneGuy**
Fixes #issue

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/common/config.h [hunk 1]
```diff
@@ -876,6 +876,10 @@ CONF_mInt32(spill_init_partition, "16");
 // The maximum size of a single log block container file, this is not a hard limit.
 // If the file size exceeds this limit, a new file will be created to store the block.
 CONF_Int64(spill_max_log_block_container_bytes, "10737418240"); // 10GB
+// The maximum size of a single spill directory, for some case the spill directory may
+// be the same with storage path. Spill will return with error when used size has exceeded
+// the limit.
+CONF_Int64(spill_max_dir_bytes, "214748364800"); // 200GB
```
Comment (silverbullet233) [inline be/src/common/config.h:882]: use can config multiple dirs for spill_local_storage_dir, should we also consider it here?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exec/spill/dir_manager.cpp [hunk 2]
```diff
@@ -76,7 +77,14 @@ Status DirManager::init() {
...
             return true;
         }));
-        _dirs.emplace_back(std::make_shared<Dir>(spill_dir_path, std::move(fs)));
+        ASSIGN_OR_RETURN(auto space_info, FileSystem::Default()->space(spill_dir_path));
+        int64_t max_dir_size = space_info.capacity;
+        for (const auto& storage_path : storage_paths) {
+            if (is_same_disk(spill_dir_path, storage_path)) {
+                max_dir_size = max_dir_size * max_dir_size_ratio;
+            }
+        }
```
Comment (silverbullet233) [inline be/src/exec/spill/dir_manager.cpp:86]: what will happen if we have multi storage_paths under the same disk?
suppose there are 2 storage_paths  and 1 spill_dir_path, they are all on the same disk,
according to the current calculation logic,
max_dir_size will be `disk_capacity* max_dir_size_ratio * max_dir_size_ratio`?

Comment (caneGuy) [reply be/src/exec/spill/dir_manager.cpp:86]: how about breaking here? should we consider multi spill dir path on the same disk?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exec/spill/dir_manager.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/exec/spill/dir_manager.cpp]: > how about breaking here? should we consider multi spill dir path on the same disk? @silverbullet233

breaking here is ok. I think it doesn't make sense to configure multi dirs under the same disk.
but if it really happens, we can:
1. add a check in init stage and return an error to user
2. accept user's configuration and make sure all dirs under the same disk share the capacity limit. e.g. there are 2 spill_dir_path under same disk, each one can use `disk_capacity * max_dir_size_ratio / 2`
3. accept user's configuration but only use the first dirs in the same disk

I think all of above are ok, which one do you prefer? @caneGuy

Comment (caneGuy) [reply be/src/exec/spill/dir_manager.cpp]: actually, i prefer return an error to user, i think one directory under the same disk is enough @silverbullet233

Comment (silverbullet233) [reply be/src/exec/spill/dir_manager.cpp]: > actually, i prefer return an error to user, i think one directory under the same disk is enough @silverbullet233

OK~

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 10 / 11 (90.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/spill/dir_manager.cpp|3|4|75.00%|[57]|
|:large_blue_circle:|src/exec/spill/dir_manager.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/spill/log_block_manager.cpp|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 22 / 27 (81.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/spill/log_block_manager.cpp|2|4|50.00%|[172, 173]|
|:large_blue_circle:|src/exec/spill/dir_manager.cpp|10|13|76.92%|[57, 58, 60]|
|:large_blue_circle:|src/exec/spill/dir_manager.h|10|10|100.00%|[]|

**Review by silverbullet233 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 22 / 27 (81.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/spill/log_block_manager.cpp|2|4|50.00%|[172, 173]|
|:large_blue_circle:|src/exec/spill/dir_manager.cpp|10|13|76.92%|[57, 58, 60]|
|:large_blue_circle:|src/exec/spill/dir_manager.h|10|10|100.00%|[]|

**Review by stdpain — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 22 / 27 (81.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/spill/log_block_manager.cpp|2|4|50.00%|[172, 173]|
|:large_blue_circle:|src/exec/spill/dir_manager.cpp|10|13|76.92%|[57, 58, 60]|
|:large_blue_circle:|src/exec/spill/dir_manager.h|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 22 / 27 (81.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/spill/log_block_manager.cpp|2|4|50.00%|[172, 173]|
|:large_blue_circle:|src/exec/spill/dir_manager.cpp|10|13|76.92%|[57, 58, 60]|
|:large_blue_circle:|src/exec/spill/dir_manager.h|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 22 / 27 (81.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/spill/log_block_manager.cpp|2|4|50.00%|[172, 173]|
|:large_blue_circle:|src/exec/spill/dir_manager.cpp|10|13|76.92%|[57, 58, 60]|
|:large_blue_circle:|src/exec/spill/dir_manager.h|10|10|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**